`CI.yml` and `environment.yml` updates to run CI on MacOS and Ubuntu

JacksonBurns opened the pull request:
This PR runs the Continuous Integration on a mac runner in addition to the Ubuntu runner, the idea being that we can catch issues that might crop up on the mac platform independently of debian linux. The results from the ubuntu runner are still considered to be the "baseline" for regression testing.

See #2355 and #2392 for issues related to running on Macs

JacksonBurns commented:
Previous CI failed due to a limitation of macos runners, see https://github.com/byrnedo/docker-reg-tool/issues/9. Updating now

JacksonBurns commented:
Switching to official Julia binaries, which may resolve the code signing problem, is blocked by the same libgfortran update that is impacting the upgrade to OpenMOPAC (i.e. potential fix is blocked by #2417). See [failed CI run](https://github.com/ReactionMechanismGenerator/RMG-Py/actions/runs/4910518427/jobs/8767760049?pr=2415) when attempting to use them.

JacksonBurns commented:
Potential bug in PATH setup for setup-julia action: https://github.com/julia-actions/setup-julia/issues/89

hwpang commented on .github/workflows/CI.yml:
```diff
@@ -104,10 +107,11 @@ jobs:
 
       # Regression Testing - Test Execution
       - name: Regression Tests - Execution
+        timeout-minutes: 60
```
What is the unit of this?

JacksonBurns replied:
`timeout-minutes: 60`
minutes

hwpang replied:
Lol can't believe I didn't see that on the first glance, thanks!

JacksonBurns commented:
Merging this PR also:
 - resolves #2392 
 - resolves #2355 
 - resolves #2290

calvinp0 commented on environment.yml:
```diff
@@ -36,8 +36,9 @@ dependencies:
+  - conda-forge::julia=1.8.5
+  - conda-forge::suitesparse
   - rmg::pydqed >=1.0.3
-  - rmg::pyjulia
```
Is pyjulia installed through `conda-forge::julia=1.8.5`?

JacksonBurns replied:
It's not, and the rmg:pyjulia does not either. The conda-recipes repo has the build script and you can see that it only provides Julia (as far as I can tell, I'm new to building conda packages). I think pyjulia is an implicit dependency of something else in the environments file? So it is added when the environment is solved?

calvinp0 replied:
So, correct if I am misunderstanding what you are saying, but we will need to keep pyjulia as it is the connection between Julia and Python. So I guess we can either keep getting from the rmg channel or from conda-forge channel.

However, the CI appears to be passing? I am a bit confused though how it is doing that when pyjulia isn't installed in the rmg_env (I am assuming here or really misunderstanding something)

JacksonBurns replied:
> So, correct if I am misunderstanding what you are saying, but we will need to keep pyjulia as it is the connection between Julia and Python. So I guess we can either keep getting from the rmg channel or from conda-forge channel.
> 
> However, the CI appears to be passing? I am a bit confused though how it is doing that when pyjulia isn't installed in the rmg_env (I am assuming here or really misunderstanding something)

My mistake - conda still ends up retrieving pyjulia from the rmg channel. See [this line in the latest CI run](https://github.com/ReactionMechanismGenerator/RMG-Py/actions/runs/4953296609/jobs/8860637395#step:4:321).

CC'ing @mjohnson - pyjulia, how does it work? The conda-recipe doesn't make sense to us.

calvinp0 replied:
So my understanding of PyJulia is that, assuming you have Julia installed and set in the PATH (or set in our conda env path for our case when rmg_env is active), it is imported in python like so:

```python
import julia
```

Then we use PyJulia for the `julia.install()` which does the whole PyCall adding (if not present) and building (unless it is M1, then it appears to have an issue with that

We can also use PyJulia to create our system images for precompiling

calvinp0 replied:
Apologies if that was already understood

mjohnson541 replied:
It is a dependency of pyrms so including pyrms in the environment.yml file will cause it to be installed as well.

calvinp0 commented on documentation/source/users/rmg/installation/anacondaDeveloper.rst:
```diff
@@ -84,6 +73,16 @@ Installation by Source Using Anaconda Environment for Unix-based Systems: Linux
+    conda create -n rmg_env
+    conda activate rmg_env
+    conda config --env --set subdir osx-64
+    conda install python=3.7
```
A bit pedantic and not really a concern or issue, but I figure we could remove the install of python=3.7 here as it will be install in the `conda env update` part of the script

JacksonBurns replied:
Yes that seems like a wasted instruction. We should try the install from scratch on a mac without it just to be sure, though.

JacksonBurns commented:
I have updated `environment.yml` to make it explicit which packages which are resolving and still depend on, and I have also added **extensive** inline documentation. I have also removed the redundant line from the developer installation instructions. The CI should pass overnight and then I will request new reviews.

JacksonBurns commented:
Tagging a lot of people for review since this PR now contains (1) a different "required" check, which requires admin access and (2) changes to the environment that will change how RMG is run on MacOS.

@hwpang and @rwest please provide general approval for this updated CI approach

@calvinp0 please check the documentation changes and ensure they match the comments from the review

@mjohnson541 please review to make sure that I did not describe any dependencies in the documented `environment.yml`

hwpang commented on .github/workflows/CI.yml:
```diff
@@ -131,7 +135,7 @@ jobs:
       - name: Upload Results as Reference
         # upload the results for scheduled CI (on main) and pushes to main
-        if: github.ref == 'refs/heads/main'
+        if: github.ref == 'refs/heads/main' && runner.os =='ubuntu'
```
I guess this means only to upload the results from the ubuntu tests since they should be the same on either platform?

JacksonBurns replied:
Yes - we treat the results on Linux as 'correct' and compare MacOS to that. Since we distribute on a Linux image for docker, I have taken that as the absolute "ground truth".

The results should never differ in the first place (hopefully).

JacksonBurns commented on .github/workflows/CI.yml:
```diff
@@ -227,14 +231,29 @@ jobs:
+  # 
+  # taken from https://github.com/orgs/community/discussions/4324#discussioncomment-3477871
+  ci-report-status:
+    runs-on: ubuntu-20.04
```
I will also add a clarification for something that confused myself, this branch is called `add-macos11-ci` but it actually runs on `macos-latest` which as of writing (May 16, 2023) is MacOS 12 Monterey

hwpang commented on documentation/source/users/rmg/installation/anacondaDeveloper.rst:
```diff
@@ -84,6 +73,15 @@ Installation by Source Using Anaconda Environment for Unix-based Systems: Linux
     cd RMG-Py
     conda env create -f environment.yml
 
+.. warning:: Apple Silicon (M1+) Users
```
Is this the same suggestion for Intel Mac?

JacksonBurns replied:
No, it only applies to M1 users (cc @calvinp0 who figured this out, thanks!)

rwest commented on environment.yml:
```diff
@@ -1,54 +1,97 @@
+  - conda-forge::openbabel >= 3
+
+# general-purpose external software tools
+  - conda-forge::julia=1.8.5
```
on second thoughts, that could be a separate pull request if you prefer (and don't mind the churn)

JacksonBurns replied:
> Dare we try `>=` and let the CI alert us if it breaks?

Move fast and break things, I say. Will try now.

> Can we put ` - conda-forge::pyjulia` here instead of the ` - rmg::pyjulia` down below? My impression from #2381 was that the answer is "yes" but maybe we can test it here?
> on second thoughts, that could be a separate pull request if you prefer (and don't mind the churn)

Every package we don't have to maintain is welcome - will try this as well.

rwest reviewed (commented):
Some suggestions that could be separate pull requests (eg. trying to change channel for pyjulia) but some fixes to the user instructions that should probably be addressed before merging.
And a question about the benefit of adding a changelog to the environment file (which is already in git with - I hope - helpful commit messages).

JacksonBurns commented on environment.yml:
```diff
@@ -1,54 +1,97 @@
+# + some other categories (see below)
+#
+# Changelog:
+# - May 15, 2023 Added this changelog, added inline documentation,
```
The way I see changelogs is that they provide a "high-level" perspective on how the file has changed over time. The git log is useful for fixing technical issues and precise problems with specific lines, but I think the changelog is better for answering questions like "why do we have some packages from this channel but other from this one" or the like.

`environment.yml` in particular has been going through substantial changes lately, and as we (hopefully) move toward using fewer and fewer in-house packages I think it will be useful to have this high level history of that activity. This is also a file that community developers will interact with a lot - such info would be particularly useful for them.

rwest commented:
Sorry - my edits introduced errors to the documentation.
<img width="1231" alt="Screenshot 2023-05-16 at 3 18 27 PM" src="https://github.com/ReactionMechanismGenerator/RMG-Py/assets/93807/ad2fee8e-bef1-4017-9c33-c4002bc2dd57">

rwest commented:
Pushed a fix for my documentation error. If/when you rebase do it with `-i --autosquash` to make it a fixup.

rwest commented:
"Job execution time - Each job in a workflow can run for up to 6 hours of execution time. If a job reaches this limit, the job is terminated and fails to complete"
"Constant Integration / build-and-test-unix (macos-latest) (pull_request) Cancelled after 360m"
Install and link julia dependencies:  5h 37m 20s
😫

rwest commented:
Your earlier run it only took about half an hour:
https://github.com/ReactionMechanismGenerator/RMG-Py/actions/runs/4987565589/jobs/8929446027

rwest commented:
I couldn't see anything wrong, so restarted the CI job. Seems to be running fine now 🤷

JacksonBurns commented:
Thanks for doing that, was going to suggest the same thing. Anecodtally, I have noticed the runners are always slower during work hours (the scheduled CI is always faster than those run during the day). I suspect that all of the containers GitHub provides are running on VMs on the same hardware, and during the day they end up fighting with one another over resources and slowing down.

rwest commented:
Sorry. The "Switch to conda-forge::pyjulia instead of the rmg channel" commit was meant to be on another branch (that's not working yet).  Will drop it on rebasing when I get a chance.

JacksonBurns commented:
> Looking good to me. Additional improvements can come in subsequent PRs.

:+1: @rwest could you update the branch protection rules to look for `ci-report-status` instead of `build-and-test-linux`, then merge?

rwest commented:
I set it to this instead:
<img width="759" alt="Screenshot 2023-05-18 at 9 33 43 AM" src="https://github.com/ReactionMechanismGenerator/RMG-Py/assets/93807/9112e4de-2dad-4b71-b59d-0815adf778c6">
which if it does  what I expect, will mean we can remove the "ci-report-status" step.
Do you think that'll work, or have I misunderstood something?

JacksonBurns commented:
@rwest that works as well - the `ci-report-status` is not needed since it is set that way, can be removed. I will roll that into a larger dead code removal PR.
